File: internal/memory/episodic.go
package memory

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"strings"
	"time"

	"github.com/caimlas/meept/pkg/sqlite"
)

const (
	// SQL for creating the episodic memories table
	createEpisodicTableSQL = `
CREATE TABLE IF NOT EXISTS episodic_memories (
    id               TEXT PRIMARY KEY,
    content          TEXT NOT NULL,
    category         TEXT NOT NULL DEFAULT 'conversation',
    metadata_json    TEXT NOT NULL DEFAULT '{}',
    created_at       TEXT NOT NULL,
    last_accessed_at TEXT NOT NULL DEFAULT '',
    version          INTEGER DEFAULT 1,
    parent_id        TEXT REFERENCES episodic_memories(id),
    is_current       INTEGER DEFAULT 1
)`

	// SQL for creating the FTS5 virtual table
	createEpisodicFTSSQL = `
CREATE VIRTUAL TABLE IF NOT EXISTS episodic_fts
USING fts5(content, category, content='episodic_memories', content_rowid='rowid')`

	// Triggers to keep FTS index in sync
	triggerEpisodicInsert = `
CREATE TRIGGER IF NOT EXISTS episodic_fts_ai AFTER INSERT ON episodic_memories BEGIN
    INSERT INTO episodic_fts(rowid, content, category)
    VALUES (new.rowid, new.content, new.category);
END`

	triggerEpisodicDelete = `
CREATE TRIGGER IF NOT EXISTS episodic_fts_ad AFTER DELETE ON episodic_memories BEGIN
    INSERT INTO episodic_fts(episodic_fts, rowid, content, category)
    VALUES ('delete', old.rowid, old.content, old.category);
END`

	triggerEpisodicUpdate = `
CREATE TRIGGER IF NOT EXISTS episodic_fts_au AFTER UPDATE ON episodic_memories BEGIN
    INSERT INTO episodic_fts(episodic_fts, rowid, content, category)
    VALUES ('delete', old.rowid, old.content, old.category);
    INSERT INTO episodic_fts(rowid, content, category)
    VALUES (new.rowid, new.content, new.category);
END`
)

// EpisodicMemory stores and retrieves conversation and interaction history.
// It uses SQLite with FTS5 for full-text search when available, falling back
// to LIKE-based queries when FTS5 is not compiled into SQLite.
type EpisodicMemory struct {
	store  *SQLiteFTSStore
	logger *slog.Logger
}

// EpisodicConfig holds configuration for episodic memory.
type EpisodicConfig struct {
	// DataDir is the directory for database files.
	DataDir string
	// Logger for operations.
	Logger *slog.Logger
}

// NewEpisodicMemory creates a new episodic memory instance.
func NewEpisodicMemory(cfg EpisodicConfig) (*EpisodicMemory, error) {
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}

	// Create the shared FTS store with episodic-specific config
	storeCfg := FTSConfig{
		TableName:     "episodic_memories",
		FTS5Table:     "episodic_fts",
		CategoryField: "category",
		DataDir:       cfg.DataDir,
		Schema:        []string{createEpisodicTableSQL, createEpisodicFTSSQL},
		Triggers:      []string{triggerEpisodicInsert, triggerEpisodicDelete, triggerEpisodicUpdate},
	}

	store, err := NewSQLiteFTSStore(storeCfg, cfg.Logger)
	if err != nil {
		return nil, fmt.Errorf("failed to create episodic store: %w", err)
	}

	return &EpisodicMemory{
		store:  store,
		logger: cfg.Logger,
	}, nil
}

// Initialize sets up the database schema and connections.
func (e *EpisodicMemory) Initialize(ctx context.Context) error {
	return e.store.Initialize(ctx)
}

// HasFTS5 returns true if FTS5 full-text search is available.
// When false, search falls back to slower LIKE-based queries.
func (e *EpisodicMemory) HasFTS5() bool {
	return e.store.HasFTS5()
}

// Store persists a new episodic memory.
// Returns the unique ID of the stored item.
func (e *EpisodicMemory) Store(ctx context.Context, content, category string, metadata map[string]any) (string, error) {
	if !e.store.HasFTS5Public() {
		e.logger.Debug("Storing without FTS5 (slower search)")
	}

	id := generateUUID()
	nowISO := time.Now().UTC().Format(time.RFC3339Nano)
	metaJSON := (&Memory{Metadata: metadata}).MetadataJSON()

	// Extract versioning fields from metadata to populate SQL columns
	var parentID sql.NullString
	var version sql.NullInt64
	var isCurrent sql.NullInt64

	if metadata != nil {
		if pid, ok := metadata["parent_id"].(string); ok && pid != "" {
			parentID = sql.NullString{String: pid, Valid: true}
		}
		if v, ok := metadata["version"]; ok {
			switch n := v.(type) {
			case int:
				version = sql.NullInt64{Int64: int64(n), Valid: true}
			case float64:
				version = sql.NullInt64{Int64: int64(n), Valid: true}
			}
		}
		if ic, ok := metadata["is_current"]; ok {
			switch n := ic.(type) {
			case int:
				isCurrent = sql.NullInt64{Int64: int64(n), Valid: true}
			case float64:
				isCurrent = sql.NullInt64{Int64: int64(n), Valid: true}
			}
		}
	}

	// Default is_current to 1 when not explicitly set
	if !isCurrent.Valid {
		isCurrent = sql.NullInt64{Int64: 1, Valid: true}
	}
	// Default version to 1 when not explicitly set
	if !version.Valid {
		version = sql.NullInt64{Int64: 1, Valid: true}
	}

	err := e.store.Store(ctx,
		`INSERT INTO episodic_memories (id, content, category, metadata_json, created_at, last_accessed_at, version, parent_id, is_current)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id, content, category, metaJSON, nowISO, nowISO, version, parentID, isCurrent,
	)
	if err != nil {
		return "", fmt.Errorf("failed to store memory: %w", err)
	}

	e.logger.Debug("Stored episodic memory", "id", id, "category", category)
	return id, nil
}

// Search finds episodic memories matching the query.
// Uses FTS5 when available, falls back to LIKE-based queries otherwise.
func (e *EpisodicMemory) Search(ctx context.Context, query string, limit int) ([]MemoryResult, error) {
	if !e.store.Initialized() {
		return nil, errors.New("episodic memory not initialized")
	}

	safeQuery := sqlite.SanitizeQuery(query)
	if safeQuery == "" {
		return e.GetRecent(ctx, limit)
	}

	hasFTS5 := e.store.HasFTS5Public()
	db := e.store.GetDB()

	var rows *sql.Rows
	var err error

	if hasFTS5 {
		// Use FTS5 for efficient full-text search
		rows, err = db.QueryContext(ctx, `
			SELECT
				m.id, m.content, m.category, m.metadata_json, m.created_at,
				f.rank
			FROM episodic_fts f
			JOIN episodic_memories m ON m.rowid = f.rowid
			WHERE episodic_fts MATCH ?
			ORDER BY f.rank
			LIMIT ?
		`, safeQuery, limit)
	} else {
		// Fallback to LIKE-based search (slower but works without FTS5)
		escapedQuery := escapeLikeWildcards(query)
		likePattern := "%" + escapedQuery + "%"
		rows, err = db.QueryContext(ctx, `
			SELECT id, content, category, metadata_json, created_at
			FROM episodic_memories
			WHERE content LIKE ? ESCAPE '\' OR category LIKE ? ESCAPE '\'
			ORDER BY created_at DESC
			LIMIT ?
		`, likePattern, likePattern, limit)
	}

	if err != nil {
		return nil, fmt.Errorf("failed to search: %w", err)
	}

	results, err := e.scanResults(rows, hasFTS5)
	rows.Close()

	if err != nil {
		return nil, err
	}

	// Update last_accessed_at asynchronously to avoid blocking.
	//nolint:gosec // goroutine outlives request context
	go func() {
		updateCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := e.updateLastAccessed(updateCtx, results); err != nil {
			e.logger.Warn("Failed to update last_accessed_at", "error", err)
		}
	}()

	return results, nil
}

// updateLastAccessed updates the last_accessed_at timestamp for retrieved memories.
func (e *EpisodicMemory) updateLastAccessed(ctx context.Context, results []MemoryResult) error {
	if len(results) == 0 {
		return nil
	}

	nowISO := time.Now().UTC().Format(time.RFC3339Nano)
	ids := make([]string, len(results))
	for i, result := range results {
		ids[i] = result.Memory.ID
	}

	placeholders := make([]string, len(ids))
	args := make([]any, len(ids)+1)
	args[0] = nowISO
	for i, id := range ids {
		placeholders[i] = "?"
		args[i+1] = id
	}

	// #nosec G201 -- all values use ? placeholders; IN clause is safe
	//nolint:gosec // parameterized query
	query := fmt.Sprintf("UPDATE episodic_memories SET last_accessed_at = ? WHERE id IN (%s)", strings.Join(placeholders, ","))

	_, err := e.store.GetDB().ExecContext(ctx, query, args...)
	return err
}

// GetRecent retrieves the most recent episodic memories.
func (e *EpisodicMemory) GetRecent(ctx context.Context, limit int) ([]MemoryResult, error) {
	if !e.store.Initialized() {
		return nil, errors.New("episodic memory not initialized")
	}

	rows, err := e.store.GetDB().QueryContext(ctx, `
		SELECT id, content, category, metadata_json, created_at
		FROM episodic_memories
		ORDER BY created_at DESC
		LIMIT ?
	`, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to get recent: %w", err)
	}
	defer rows.Close()

	return e.scanResults(rows, false)
}

// GetByCategory retrieves memories filtered to a specific category.
func (e *EpisodicMemory) GetByCategory(ctx context.Context, category string, limit int) ([]MemoryResult, error) {
	rows, err := e.store.GetDB().QueryContext(ctx, `
		SELECT id, content, category, metadata_json, created_at
		FROM episodic_memories
		WHERE category = ?
		ORDER BY created_at DESC
		LIMIT ?
	`, category, limit)
	if err != nil {
		return nil, fmt.Errorf("failed to get by category: %w", err)
	}
	defer rows.Close()

	return e.scanResults(rows, false)
}

// GetByTimeRange retrieves memories within a time range.
func (e *EpisodicMemory) GetByTimeRange(ctx context.Context, start, end time.Time, limit int) ([]MemoryResult, error) {
	rows, err := e.store.GetDB().QueryContext(ctx, `
		SELECT id, content, category, metadata_json, created_at
		FROM episodic_memories
		WHERE created_at >= ? AND created_at <= ?
		ORDER BY created_at DESC
		LIMIT ?
	`, start.UTC().Format(time.RFC3339Nano), end.UTC().Format(time.RFC3339Nano), limit)
	if err != nil {
		return nil, fmt.Errorf("failed to get by time range: %w", err)
	}
	defer rows.Close()

	return e.scanResults(rows, false)
}

// GetOldMemories retrieves memories older than the given time.
func (e *EpisodicMemory) GetOldMemories(ctx context.Context, olderThan time.Time, limit int) ([]MemoryResult, error) {
	rows, err := e.store.GetDB().QueryContext(ctx, `
		SELECT id, content, category, metadata_json, created_at
		FROM episodic_memories
		WHERE created_at < ?
		ORDER BY created_at ASC
		LIMIT ?
	`, olderThan.UTC().Format(time.RFC3339Nano), limit)
	if err != nil {
		return nil, fmt.Errorf("failed to get old memories: %w", err)
	}
	defer rows.Close()

	return e.scanResults(rows, false)
}

// Delete removes a memory by ID.
func (e *EpisodicMemory) Delete(ctx context.Context, id string) error {
	return e.store.Delete(ctx, "DELETE FROM episodic_memories WHERE id = ?", id)
}

// GetByID retrieves a single memory by its ID.
// Returns nil, nil if the memory is not found.
func (e *EpisodicMemory) GetByID(ctx context.Context, id string) (*MemoryResult, error) {
	if !e.store.Initialized() {
		return nil, errors.New("episodic memory not initialized")
	}

	row := e.store.GetDB().QueryRowxContext(ctx, `
		SELECT id, content, category, metadata_json, created_at
		FROM episodic_memories
		WHERE id = ?
	`, id)

	var memID, content, category, metaJSON, createdAtStr string
	err := row.Scan(&memID, &content, &category, &metaJSON, &createdAtStr)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, fmt.Errorf("failed to get memory by ID: %w", err)
	}

	createdAt, _ := time.Parse(time.RFC3339Nano, createdAtStr)
	mem := Memory{
		ID:        memID,
		Content:   content,
		Type:      MemoryTypeEpisodic,
		Category:  category,
		Metadata:  ParseMetadata(metaJSON),
		CreatedAt: createdAt,
	}

	return &MemoryResult{
		Memory: mem,
		Source: "episodic",
	}, nil
}

// DeleteByIDs removes multiple memories by ID.
func (e *EpisodicMemory) DeleteByIDs(ctx context.Context, ids []string) (int, error) {
	return e.store.DeleteByIDs(ctx, "episodic_memories", ids)
}

// Count returns the total number of episodic memories.
func (e *EpisodicMemory) Count(ctx context.Context) (int, error) {
	return e.store.Count(ctx, "episodic_memories")
}

// GetOldestTimestamp returns the created_at of the oldest memory.
func (e *EpisodicMemory) GetOldestTimestamp(ctx context.Context) (*time.Time, error) {
	return e.store.GetOldestTimestamp(ctx, "episodic_memories")
}

// GetNewestTimestamp returns the created_at of the newest memory.
func (e *EpisodicMemory) GetNewestTimestamp(ctx context.Context) (*time.Time, error) {
	return e.store.GetNewestTimestamp(ctx, "episodic_memories")
}

// Close releases all resources.
func (e *EpisodicMemory) Close() error {
	return e.store.Close()
}

// scanResults scans database rows into MemoryResult slice.
// Delegates to the shared SQLiteFTSStore.ScanResults implementation.
func (e *EpisodicMemory) scanResults(rows *sql.Rows, hasRank bool) ([]MemoryResult, error) {
	return e.store.ScanResults(rows, hasRank, ScanRowConfig{
		MemoryType: MemoryTypeEpisodic,
		SourceFmt:  "episodic",
	})
}

// Ensure EpisodicMemory implements io.Closer
var _ io.Closer = (*EpisodicMemory)(nil)

File: archive/python-legacy/meept/memory/episodic.py
"""Episodic memory subsystem -- conversation and interaction history.

Attempts to use ``memu`` for vector-backed episodic storage.  When ``memu``
is not installed the module falls back to a fully-functional SQLite + FTS5
implementation so the rest of meept keeps working without external
dependencies.
"""

from __future__ import annotations

import json
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import aiosqlite

from meept.models.memory_types import MemoryItem, MemoryResult, MemoryType

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Probe for memu availability
# ---------------------------------------------------------------------------

_MEMU_AVAILABLE: bool
try:
    import memu  # type: ignore[import-untyped]

    _MEMU_AVAILABLE = True
    logger.debug("memu library detected -- using memu for episodic memory.")
except ImportError:
    _MEMU_AVAILABLE = False
    logger.debug("memu not installed -- episodic memory will use SQLite/FTS5 fallback.")


# ---------------------------------------------------------------------------
# SQLite FTS5 fallback
# ---------------------------------------------------------------------------

_CREATE_TABLE_SQL = """\
CREATE TABLE IF NOT EXISTS episodic_memories (
    id            TEXT PRIMARY KEY,
    content       TEXT NOT NULL,
    category      TEXT NOT NULL DEFAULT 'conversation',
    metadata_json TEXT NOT NULL DEFAULT '{}',
    created_at    TEXT NOT NULL,
    embedding_text TEXT NOT NULL DEFAULT ''
);
"""

_CREATE_FTS_SQL = """\
CREATE VIRTUAL TABLE IF NOT EXISTS episodic_fts
USING fts5(content, category, embedding_text, content='episodic_memories', content_rowid='rowid');
"""

# Triggers that keep the FTS index in sync with the content table.
_TRIGGER_INSERT = """\
CREATE TRIGGER IF NOT EXISTS episodic_fts_ai AFTER INSERT ON episodic_memories BEGIN
    INSERT INTO episodic_fts(rowid, content, category, embedding_text)
    VALUES (new.rowid, new.content, new.category, new.embedding_text);
END;
"""

_TRIGGER_DELETE = """\
CREATE TRIGGER IF NOT EXISTS episodic_fts_ad AFTER DELETE ON episodic_memories BEGIN
    INSERT INTO episodic_fts(episodic_fts, rowid, content, category, embedding_text)
    VALUES ('delete', old.rowid, old.content, old.category, old.embedding_text);
END;
"""

_TRIGGER_UPDATE = """\
CREATE TRIGGER IF NOT EXISTS episodic_fts_au AFTER UPDATE ON episodic_memories BEGIN
    INSERT INTO episodic_fts(episodic_fts, rowid, content, category, embedding_text)
    VALUES ('delete', old.rowid, old.content, old.category, old.embedding_text);
    INSERT INTO episodic_fts(rowid, content, category, embedding_text)
    VALUES (new.rowid, new.content, new.category, new.embedding_text);
END;
"""


# ---------------------------------------------------------------------------
# Public class
# ---------------------------------------------------------------------------


class EpisodicMemory:
    """Stores and retrieves episodic (conversation) memories.

    If the ``memu`` library is installed it is used for vector-similarity
    search.  Otherwise an SQLite database with FTS5 full-text search provides
    equivalent functionality with keyword matching instead of embeddings.
    """

    def __init__(self) -> None:
        self._db: aiosqlite.Connection | None = None
        self._data_dir: Path | None = None
        self._use_memu: bool = False
        self._memu_store: Any = None
        self._llm_client: Any = None
        self._initialized: bool = False

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def initialize(self, data_dir: Path, llm_client: Any = None) -> None:
        """Set up the episodic memory backend.

        Parameters
        ----------
        data_dir:
            Directory in which database files are stored.
        llm_client:
            Optional :class:`~meept.llm.client.LLMClient` used for smart
            retrieval when memu is not available.
        """
        self._data_dir = data_dir
        self._llm_client = llm_client
        data_dir.mkdir(parents=True, exist_ok=True)

        if _MEMU_AVAILABLE:
            try:
                self._memu_store = memu.MemoryStore(str(data_dir / "episodic_memu"))
                self._use_memu = True
                logger.info("Episodic memory initialised with memu backend.")
            except Exception:
                logger.warning(
                    "Failed to initialise memu store -- falling back to SQLite.",
                    exc_info=True,
                )
                self._use_memu = False

        # Always open the SQLite database -- it serves as fallback *and* as
        # the canonical metadata / category / archive store even when memu is
        # the primary search backend.
        db_path = data_dir / "episodic.db"
        self._db = await aiosqlite.connect(str(db_path))
        await self._db.execute("PRAGMA journal_mode=WAL;")
        await self._db.execute("PRAGMA foreign_keys=ON;")
        await self._db.execute(_CREATE_TABLE_SQL)
        await self._db.execute(_CREATE_FTS_SQL)
        await self._db.execute(_TRIGGER_INSERT)
        await self._db.execute(_TRIGGER_DELETE)
        await self._db.execute(_TRIGGER_UPDATE)
        await self._db.commit()

        self._initialized = True
        logger.info(
            "Episodic memory ready (backend=%s, db=%s).",
            "memu" if self._use_memu else "sqlite/fts5",
            db_path,
        )

    # ------------------------------------------------------------------
    # Store
    # ------------------------------------------------------------------

    async def store(
        self,
        content: str,
        category: str = "conversation",
        metadata: dict | None = None,
    ) -> str:
        """Persist a new episodic memory.

        Returns the unique id of the stored item.
        """
        self._ensure_initialized()
        assert self._db is not None

        item_id = uuid.uuid4().hex
        now_iso = datetime.now(timezone.utc).isoformat()
        meta_json = json.dumps(metadata or {})
        # embedding_text combines content + category for better FTS coverage.
        embedding_text = f"{category}: {content}"

        await self._db.execute(
            """
            INSERT INTO episodic_memories (id, content, category, metadata_json, created_at, embedding_text)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (item_id, content, category, meta_json, now_iso, embedding_text),
        )
        try:
            await self._db.commit()
        except Exception:
            logger.error("Failed to commit episodic memory %s", item_id, exc_info=True)
            raise

        # Also index in memu when available.
        if self._use_memu and self._memu_store is not None:
            try:
                self._memu_store.add(
                    item_id,
                    content,
                    metadata={"category": category, **(metadata or {})},
                )
            except Exception:
                logger.warning("memu indexing failed for %s", item_id, exc_info=True)

        logger.debug("Stored episodic memory %s (category=%s).", item_id, category)
        return item_id

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------

    async def search(self, query: str, limit: int = 10) -> list[MemoryResult]:
        """Search episodic memories by relevance.

        Uses memu vector search when available, otherwise falls back to FTS5.
        """
        self._ensure_initialized()

        if self._use_memu and self._memu_store is not None:
            return await self._search_memu(query, limit)
        return await self._search_fts(query, limit)

    async def get_recent(self, limit: int = 20) -> list[MemoryResult]:
        """Retrieve the most recent episodic memories by creation time."""
        self._ensure_initialized()
        assert self._db is not None

        rows = await self._fetch_all(
            """
            SELECT id, content, category, metadata_json, created_at
            FROM episodic_memories
            ORDER BY created_at DESC
            LIMIT ?
            """,
            (limit,),
        )
        return [self._row_to_result(r, relevance=0.0) for r in rows]

    async def get_by_category(
        self, category: str, limit: int = 20
    ) -> list[MemoryResult]:
        """Retrieve memories filtered to a specific category."""
        self._ensure_initialized()
        assert self._db is not None

        rows = await self._fetch_all(
            """
            SELECT id, content, category, metadata_json, created_at
            FROM episodic_memories
            WHERE category = ?
            ORDER BY created_at DESC
            LIMIT ?
            """,
            (category, limit),
        )
        return [self._row_to_result(r, relevance=0.0) for r in rows]

    # ------------------------------------------------------------------
    # Stats helper (used by consolidation)
    # ------------------------------------------------------------------

    async def count(self) -> int:
        """Return total number of episodic memories."""
        self._ensure_initialized()
        assert self._db is not None
        async with self._db.execute("SELECT COUNT(*) FROM episodic_memories") as cur:
            row = await cur.fetchone()
            return row[0] if row else 0

    async def get_oldest_timestamp(self) -> datetime | None:
        """Return the created_at of the oldest memory, or ``None``."""
        self._ensure_initialized()
        assert self._db is not None
        async with self._db.execute(
            "SELECT MIN(created_at) FROM episodic_memories"
        ) as cur:
            row = await cur.fetchone()
            if row and row[0]:
                return datetime.fromisoformat(row[0])
        return None

    async def get_newest_timestamp(self) -> datetime | None:
        """Return the created_at of the newest memory, or ``None``."""
        self._ensure_initialized()
        assert self._db is not None
        async with self._db.execute(
            "SELECT MAX(created_at) FROM episodic_memories"
        ) as cur:
            row = await cur.fetchone()
            if row and row[0]:
                return datetime.fromisoformat(row[0])
        return None

    async def get_old_memories(
        self, older_than: datetime, limit: int = 200
    ) -> list[MemoryResult]:
        """Return memories created before *older_than* (for consolidation)."""
        self._ensure_initialized()
        assert self._db is not None

        rows = await self._fetch_all(
            """
            SELECT id, content, category, metadata_json, created_at
            FROM episodic_memories
            WHERE created_at < ?
            ORDER BY created_at ASC
            LIMIT ?
            """,
            (older_than.isoformat(), limit),
        )
        return [self._row_to_result(r, relevance=0.0) for r in rows]

    async def delete_by_ids(self, ids: list[str]) -> int:
        """Delete memories by their ids. Returns count of deleted rows."""
        self._ensure_initialized()
        assert self._db is not None

        if not ids:
            return 0

        placeholders = ",".join("?" for _ in ids)
        async with self._db.execute(
            f"DELETE FROM episodic_memories WHERE id IN ({placeholders})",  # noqa: S608
            ids,
        ) as cur:
            deleted = cur.rowcount
        try:
            await self._db.commit()
        except Exception:
            logger.error("Failed to commit deletion of %d episodic memories", len(ids), exc_info=True)
            raise

        # Also remove from memu index.
        if self._use_memu and self._memu_store is not None:
            for item_id in ids:
                try:
                    self._memu_store.remove(item_id)
                except Exception:
                    pass

        return deleted

    # ------------------------------------------------------------------
    # Cleanup
    # ------------------------------------------------------------------

    async def close(self) -> None:
        """Release resources."""
        if self._db is not None:
            await self._db.close()
            self._db = None
        if self._use_memu and self._memu_store is not None:
            try:
                self._memu_store.close()
            except Exception:
                pass
            self._memu_store = None
        self._initialized = False
        logger.info("Episodic memory closed.")

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _ensure_initialized(self) -> None:
        if not self._initialized:
            raise RuntimeError(
                "EpisodicMemory has not been initialised. "
                "Call ``await initialize(data_dir)`` first."
            )

    async def _fetch_all(
        self, sql: str, params: tuple = ()
    ) -> list[tuple]:
        """Execute a query and return all rows."""
        assert self._db is not None
        async with self._db.execute(sql, params) as cursor:
            return await cursor.fetchall()

    def _row_to_result(
        self, row: tuple, relevance: float = 0.0
    ) -> MemoryResult:
        """Convert a raw SQLite row to a ``MemoryResult``."""
        item_id, content, category, meta_json, created_at_str = row[:5]
        try:
            metadata = json.loads(meta_json)
        except (json.JSONDecodeError, TypeError):
            metadata = {}
        created_at = datetime.fromisoformat(created_at_str)

        item = MemoryItem(
            id=item_id,
            content=content,
            memory_type=MemoryType.EPISODIC,
            category=category,
            metadata=metadata,
            created_at=created_at,
        )
        return MemoryResult(item=item, relevance_score=relevance, source="episodic")

    async def _search_fts(self, query: str, limit: int) -> list[MemoryResult]:
        """Full-text search using SQLite FTS5."""
        assert self._db is not None

        # FTS5 match query -- escape double-quotes in the user query to avoid
        # syntax errors, and wrap each token in quotes for a safe prefix search.
        safe_query = _sanitise_fts_query(query)
        if not safe_query:
            # If the query sanitises to nothing, fall back to recent memories.
            return await self.get_recent(limit)

        rows = await self._fetch_all(
            """
            SELECT
                m.id, m.content, m.category, m.metadata_json, m.created_at,
                rank
            FROM episodic_fts f
            JOIN episodic_memories m ON m.rowid = f.rowid
            WHERE episodic_fts MATCH ?
            ORDER BY rank
            LIMIT ?
            """,
            (safe_query, limit),
        )

        results: list[MemoryResult] = []
        for r in rows:
            # FTS5 rank is negative (lower = better); normalise to [0, 1].
            raw_rank = r[5] if len(r) > 5 else 0.0
            score = _normalise_fts_rank(raw_rank)
            results.append(self._row_to_result(r, relevance=score))
        return results

    async def _search_memu(self, query: str, limit: int) -> list[MemoryResult]:
        """Vector-similarity search via memu."""
        assert self._memu_store is not None
        assert self._db is not None

        try:
            hits = self._memu_store.search(query, top_k=limit)
        except Exception:
            logger.warning("memu search failed -- falling back to FTS.", exc_info=True)
            return await self._search_fts(query, limit)

        results: list[MemoryResult] = []
        for hit in hits:
            # memu returns objects with .id, .score, .metadata
            hit_id = getattr(hit, "id", None) or (hit.get("id") if isinstance(hit, dict) else None)
            hit_score = getattr(hit, "score", 0.0) if not isinstance(hit, dict) else hit.get("score", 0.0)

            if hit_id is None:
                continue

            # Fetch full row from SQLite for metadata.
            rows = await self._fetch_all(
                """
                SELECT id, content, category, metadata_json, created_at
                FROM episodic_memories
                WHERE id = ?
                """,
                (str(hit_id),),
            )
            if rows:
                results.append(self._row_to_result(rows[0], relevance=float(hit_score)))
        return results


# ---------------------------------------------------------------------------
# Module-level helpers
# ---------------------------------------------------------------------------


def _sanitise_fts_query(raw: str) -> str:
    """Turn a user-provided string into a safe FTS5 MATCH expression.

    Each whitespace-separated token is double-quoted so special FTS5
    characters (``*``, ``:``, ``(``, etc.) are treated as literals.
    Tokens are joined with implicit AND.
    """
    tokens = raw.split()
    safe_tokens: list[str] = []
    for t in tokens:
        # Remove characters that break FTS5 even inside double-quotes.
        cleaned = t.replace('"', "").strip()
        if cleaned:
            safe_tokens.append(f'"{cleaned}"')
    return " ".join(safe_tokens)


def _normalise_fts_rank(rank: float) -> float:
    """Map an FTS5 ``rank`` value to ``[0.0, 1.0]``.

    FTS5 rank values are negative (more negative = better match).  We map
    via ``score = 1 / (1 + abs(rank))`` so that more-negative ranks produce
    higher scores (closer to 1.0).
    """
    if rank >= 0:
        return 0.0
    return 1.0 / (1.0 + abs(rank))
